### DISCUSSION TITLE
How do you access the web GUI remotely?

### DISCUSSION THREAD

**Discussion by jl-678** [FoldingAtHome/fah-web-client-bastet#136] [General]
My FAH Debian Buster server is headless. With v7, I could access the basic client GUI remotely and while its functionality was limited, it was something. I am trying to determine how to do something similar with v8.1.11 and am currently running the client with the following command.

`/usr/bin/fah-client --allow=<home_ip_range>/24 --http-addresses=<IP of local machine>:7396 --log=/var/log/fah-client/log.txt --log-rotate-dir=/var/log/fah-client/`

Unfortunately, I get a 404 error when I go to http://<IP of local machine>:7396, and https://app.foldingathome.org/ is stuck on "connecting" because it cannot be run locally.

Am I missing something with the above configuration parameters? Is there something else I can do to enable this use case?

Thank you.

**Comment by kbernhagen** [comment]
You need `--http-addresses=0.0.0.0:7396`

You should probably always include `127.0.0.1` in allow.
You need `--deny=0/0` for allow to work.
The defaults are effectively allow any, deny none.

**Comment by kbernhagen** [comment]
If you don't want to use a localhost client as a trampoline via peers,
you can load https://app.foldingathome.org/ and type Control-Shift-L to change the client.
This is stored in the browser database for the site.

**Comment by jl-678** [comment]
Okay, so I ended up with this:

`/usr/bin/fah-client --allow=127.0.0.1 --deny=0/0 --http-addresses=0.0.0.0:7396 --log=/var/log/fah-client/log.txt --log-rotate-dir=/var/log/fah-client/`

Then I go to the client and do the control-shift-L thing, and it sits there saying "Connecting." (The screen is static.) This is on a Debian 10 system. I also have access to an Ubuntu system if you want me try that.

What am I missing?

Edit:
I see these exceptions, but I think that they relate to my lack of a GPU.  However, I wanted to include these lines in case they relate.
```
21:41:59:E :Exception: Failed to open dynamic library 'libcuda.so': libcuda.so: cannot open shared object file: No such file or directory
21:41:59:E :Exception: clGetPlatformIDs() returned -1001
```

**Comment by kbernhagen** [comment]
You need something like `--allow="127.0.0.1 10.0.0.0/8 192.168.0.0/16 172.16.0.0/12 169.254.0.0/16"`

You can remove the addresses that don't apply.

You should add your changes to a `config.xml` in the working directory for the client.

I believe by default it is a symlink to `/etc/fahclient/config.xml`.

You are using the systemd service as installed by the deb package?

**Comment by jl-678** [comment]
I will test that out.  I am currently not using a config.xml or systemd instead am running from a simple shell script including the above command.  Once I get it running correctly, I will update the config.xml accordingly.  However, I find the script easier as I can watch the output.

Edit:
It might be good to log failed access like this to help with troubleshooting.  Maybe I up the log level on the CLI to get that?

**Comment by kbernhagen** [comment]
Example `config.xml`

```
<config>
  <http-addresses v='0.0.0.0:7396'/>
  <allow v='127.0.0.1 10.0.0.0/8 192.168.0.0/16 172.16.0.0/12 169.254.0.0/16'/>
  <deny v='0/0'/>
</config>
```

**Comment by jl-678** [comment]
Okay, that matches v7. I assume that it will work equally well on the CLI for testing. Right? Meaning passing the data in directly to fah-client.

**Comment by kbernhagen** [comment]
If you have the debug build, don't go above verbosity 3.

**Comment by jl-678** [comment]
I am running standard. (e.g. non-debug.) I can use that version if you prefer.

**Comment by kbernhagen** [comment]
The command line overrides what is found in config.xml, if it exists. Same otherwise.

Anything exposed in Web Control is stored in client.db. For those values, config.xml is just first-run defaults.

**Comment by kbernhagen** [comment]
Try release with higher verbosity first.
I can't remember what is logged where at what verbosity.

**Comment by jl-678** [comment]
Question where is client db stored? I wonder if it has legacy data from v7 that is causing problems. I assume that it is a sqlite DB. Correct?

Here is the latest with the same issue:
`/usr/bin/fah-client --allow='127.0.0.1 10.25.0.0/24' --deny=0/0 --http-addresses=0.0.0.0:7396 --log=/var/log/fah-client/log.txt --log-rotate-dir=/var/log/fah-client/ --verbosity=5`

The logs look okay as far as I can tell. There are no obvious error that appear beside the aforementioned libcuda stuff, and 5 seems to be the highest verbosity. (I tried 20 and saw nothing more.)

Should I switch to the debug version?

**Comment by kbernhagen** [comment]
`client.db` is created in the current working directory of the running client.
v7 uses a different database `./work/client.db`.
There can be no conflict.

Yes, sqlite3. Please do not try to edit it.

**Comment by kbernhagen** [comment]
What you have looks correct to me.

Assuming you're trying to connect from subnet `10.25.0.0/24`.

Please try debug build verbosity 3.

**Comment by kbernhagen** [comment]
Are you running the client inside a container?

**Comment by jl-678** [comment]
Hmmm, I have two systems running this debian 10 and Ubuntu, and both show the same behavior. I installed the debug version on the Ubuntu host and see nothing that looks different. (level 5) One aside is that all these systems are LXC containers, not sure if that matters, but I have never encountered issues with that in the past and v7 had no issues.

I have also tried the shift-ctrl-L thing on another laptop with the same issue.

The IP is a fake one. But yes, both browsers are trying to connect from a host in that subnet.

**Comment by kbernhagen** [comment]
I think the LXC containers do not have bridged networking by default.
I cannot help you configure them, due to ignorance.

**Comment by jl-678** [comment]
I have never had a network issue. LXC containers appear as normal hosts. It has a full IP address like any physical host.  This is actually why I like LXC containers they behave like actual hosts without all the bridging overhead that comes with Docker.

**Comment by kbernhagen** [comment]
All I can suggest is allowing all non-routable addresses.
Also note, the client only supports IPv4 addresses.

**Comment by jl-678** [comment]
Tell you what, I have an unused VM sitting around and will try that and see what happens. Standby.

All non-routable aka 0/0?

**Comment by kbernhagen** [comment]
All non-routable IPv4:
```
<allow v='127.0.0.1 10.0.0.0/8 192.168.0.0/16 172.16.0.0/12 169.254.0.0/16'/>
```
A.k.a., private network addresses.

**Comment by jl-678** [comment]
Frustrating....I have switched to a VM to avoid the unlikely possibility of routing issues.  Here is the script content:

```
#!/bin/bash
/usr/bin/fah-client --allow='127.0.0.1 10.0.0.0/8 192.168.0.0/16 172.16.0.0/12 169.254.0.0/16' --deny=0/0 --http-addresses=0.0.0.0:7396 --log=/var/log/fah-client/log.txt --log-rotate-dir=/var/log/fah-client/ --verbosity=3
```
I still see nothing troubling in the log file, and I see the same issue. Is there anything else I can do to help troubleshoot?

This feels like some kind of network communication issue since there is nothing in the debug stream. I have verified that I can ping all systems.....

**Comment by kbernhagen** [comment]
Please try without allow/deny.

**Comment by jl-678** [comment]
`/usr/bin/fah-client --http-addresses=0.0.0.0:7396 --log=/var/log/fah-client/log.txt --log-rotate-dir=/var/log/fah-client/ --verbosity=10`

Same issue. I am sorry that this is so bothersome....

Oh and I see nothing troubling in the verbosity=10 except the basic config, the db creation and lots of event callbacks.

**Comment by jl-678** [comment]
Maybe I should go put all this stuff in the config file? Not sure if that matters.

**Comment by kbernhagen** [comment]
Is IPv4 enabled?
Is it set as preferred over IPv6? (I don't know how to do so)

**Comment by kbernhagen** [comment]
I might matter, but shouldn't.
If it does it's a bug in [cbang](https://github.com/CauldronDevelopmentLLC/cbang).

**Comment by jl-678** [comment]
So here is an interesting datapoint, I go to http:<FAH_IP>:7396. I see the connection come in.  I am clearly able to reach FAH on my browser which rules out local connectivity.

```
22:51:03:I1:CON2:< <IP Address>:64037 GET /favicon.ico HTTP/1.1
22:51:03:D5:CON2:Host: <FAH_IP>:7396
22:51:03:D5:Connection: keep-alive
22:51:03:D5:User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36
22:51:03:D5:DNT: 1
22:51:03:D5:Accept: image/avif,image/webp,image/apng,image/svg+xml,image/*,*/*;q=0.8
22:51:03:D5:Referer: http://<FAH_IP>:7396/
22:51:03:D5:Accept-Encoding: gzip, deflate
22:51:03:D5:Accept-Language: en-US,en;q=0.9
22:51:03:D5:
22:51:03:D6:CON2:
22:51:03:I1:REQ2:> HTTP/1.1 404 HTTP_NOT_FOUND
22:51:03:D5:REQ2:Content-Type: text/html
22:51:03:D5:Connection: close
```

**Comment by kbernhagen** [comment]
You cannot connect to the client directly via browser, unless you also set `web-root` to point to a built static copy of web control dist.

If you have `npm`, I can tell you how.

**Comment by jl-678** [comment]
If you are referring to the NPM client stuff, I tried that a bit ago too and had the same issue. I am open for trying to make this work.

Edit1:
I was thinking that the log would show that the IPv4 network seems to be working as I can access the FAH server via web browser.

Edit2:
I am referring to the fah-web-client-bastet-main stuff. I have that installed and tried running it on the LXC Ubuntu host. It has no issue running, but seems to have the same problem connecting.

**Comment by kbernhagen** [comment]
Similar to the readme at https://github.com/FoldingAtHome/fah-web-client-bastet:

```
cd /path/to/client/data-dir
git clone https://github.com/foldingathome/fah-web-client-bastet
cd fah-web-client-bastet
npm install
npm run build
```
Then, in `config.xml`, add
```
  <allowed-origins v='https://app.foldingathome.org http://localhost:7396 http://yourmonitoringcomputer.local:7396'/>
  <web-root v='./fah-web-client-bastet/dist'/>
```

Restart client whenever you make networking changes.

You can get `npm` easily if you use [homebrew](https://brew.sh)
```
brew install npm
```

**Comment by kbernhagen** [comment]
Note, for https://app.foldingathome.org to access remote peers, you need to allow insecure content for that site in your browser.

**Comment by kbernhagen** [comment]
Firefox will only allow insecure access to localhost. Safari will not allow insecure access at all.

**Comment by jl-678** [comment]
On the NPM stuff, I already most of that and so let me and following the directions [here](https://github.com/foldingathome/fah-web-client-bastet), I run it with this:

`npm run dev -- --host`

I will explore adding the config.xml stuff to that same host.

**Comment by kbernhagen** [comment]
No. Use npm run build 
The client has a built in web server and will serve the built static app files.

**Comment by kbernhagen** [comment]
Access with http://myserver:7396

**Comment by jl-678** [comment]
See the logs below. It still shows "Connecting"

```
00:13:03:I2:  <!-- Client -->
00:13:03:I2:  <allowed-origins v='https://app.foldingathome.org http://localhost:7396 http://<fah_ip>:7396'/>
00:13:03:I2:  <web-root v='./fah-web-client-bastet/dist'/>
00:13:03:I2:
00:13:03:I2:  <!-- HTTP Server -->
00:13:03:I2:  <allow v='127.0.0.1 10.0.0.0/8 192.168.0.0/16 172.16.0.0/12 169.254.0.0/16'/>
00:13:03:I2:  <deny>
00:13:03:I2:    0/0
00:13:03:I2:  </deny>
00:13:03:I2:  <http-addresses v='0.0.0.0:7396'/>
00:13:03:I2:</config>
00:13:03:I1:Opening Database
00:13:03:I1:Listening for HTTP on 0.0.0.0:7396
00:13:03:I3:id = XsXcL5bOTGfSrDkHj8Dh7hswaiOrqFElLDo6yqU/3nE=
00:13:03:I3:Loaded 0 wus.
00:13:03:D1::Remaining CPUs: 1, Remaining GPUs: 0, Active WUs: 0
00:13:03:E :Exception: Failed to open dynamic library 'libcuda.so': libcuda.so: cannot open shared object file: No such file or directory
00:13:03:E :       At: src/cbang/os/DynamicLibrary.cpp:85:cb::DynamicLibrary::DynamicLibrary()
00:13:03:E :Caught at: src/fah/client/GPUResources.cpp:68:{anonymous}::match()
00:13:03:E :Exception: Failed to open dynamic library 'libOpenCL.so': libOpenCL.so: cannot open shared object file: No such file or directory
00:13:03:E :       At: src/cbang/os/DynamicLibrary.cpp:85:cb::DynamicLibrary::DynamicLibrary()
00:13:03:E :Caught at: src/fah/client/GPUResources.cpp:68:{anonymous}::match()
00:13:06:I1:CON1:< <laptop IP>:65114 GET / HTTP/1.1
00:13:06:I1:CON2:< <laptop IP>:65115 GET /assets/index.e05fe103.js HTTP/1.1
00:13:06:I1:CON1:< <laptop IP>:65114 GET /css/font-awesome.min.css HTTP/1.1
00:13:06:I1:CON3:< <laptop IP>:65116 GET /assets/index.60f7f7bb.css HTTP/1.1
00:13:06:I1:CON2:< <laptop IP>:65115 GET /fonts/fontawesome-webfont.woff2?v=4.7.0 HTTP/1.1
00:13:06:I1:CON2:< <laptop IP>:65115 GET /favicon.ico HTTP/1.1
```

**Comment by kbernhagen** [comment]
In your web browser, open the JavaScript console and look for errors. You might need to set the console to verbose. 

Oops. I forgot you currently need to patch client.js before the Npm build. When I’m back at my computer I will post the change. 

Meanwhile, you should be able to use https://app.foldingathome.org if you go to site settings and allow insecure content.

**Comment by kbernhagen** [comment]
You can currently see my changes at https://github.com/kbernhagen/fah-web-client-bastet/commit/f495517cda8b621fc3f75231874f5f8957885d7b

**Comment by kbernhagen** [comment]
Instead of sanctuary.local you would use your server hostname or ip.

**Comment by kbernhagen** [comment]
Actually, when it says connecting, you can possibly use Control-shift-L

**Comment by kbernhagen** [comment]
That’s not as good if you want to direct connect to multiple servers.

But once you have one client connecting, you can add others as peers.

**Comment by jl-678** [comment]
It works!!!!! I think that maybe part of the problem with the test was that going to <IP>:7396 results in the system trying to connect to 127.0.0.1 (when hitting ctr-shift-l you see this.)

Not sure if the js fixes you mentioned did anything, but I added those too.

**Comment by jl-678** [comment]
Thank you SOOOOO much! I really appreciate your patience.

Question for you. Can I help you on this project with troubleshooting or testing on one of my other systems?

BTW, the successful system is the LXC container running Ubuntu.

**Comment by jl-678** [comment]
Question: I see a bunch of this in the work unit log.  Should I be concerned?

```
BLE: {"error":"No appropriate assignment"}
01:35:21:I1::WU3:Retry #3 in 8 secs
01:35:29:I1::WU3:Requesting WU assignment
01:35:29:I1:OUT9:> POST https://assign4.foldingathome.org/api/assign HTTP/1.1
01:35:30:I1:OUT9:< assign4.foldingathome.org:443 HTTP/1.1 503 HTTP_SERVICE_UNAVAILABLE
01:35:30:E ::WU3:HTTP_SERVICE_UNAVAIL
```

**Comment by kbernhagen** [comment]
Yeah! You're now an expert.
Maybe you can help others on foldingforum.org, as time permits?

I haven't heard about anyone testing high thread-count servers.
If you have anything with more than 64 threads, please report your experiences on the forum. Any Windows machines with more than 32 threads too.

Please continue to report bugs and feature requests on GitHub.
Please also check closed issues before creating a new issue.

I wouldn't worry about that log output as long as you get WU assignments.

**Comment by kbernhagen** [comment]
Did you also get https://app.foldingathome.org working?

**Comment by jl-678** [comment]
No, that did not work. I could only get there via http://ip:7396.

I am still concerned about getting jobs. I will circle back in that.

**Comment by kbernhagen** [comment]
Yea, not getting assignments is a different issue.
Please ask on https://foldingforum.org
Folks there will be familiar with the myriad ways that can break.

**Comment by jl-678** [comment]
Okay, I just wanted to be sure that it was not related to the beta. I do have a couple of questions about this:

- There is a Beta checkbox under project settings. What does that do and why would I check it?
- There is a drop down called "cause" which is unpopulated. Is something supposed to be in there?

This is particularly curious because in v7, I had the ability to choose a cause and do not seem to be able to do that in the v8 gui.

**Comment by kbernhagen** [comment]
The beta checkbox is for beta WUs.
Do not use it unless you have been told to.

The causes issue I answered on the forum.

**Comment by jl-678** [comment]
Okay, thx. I assume that you want to talk API and WUs in the forum and so I will follow-up there.

**Comment by kbernhagen** [comment]
@jcoffland please consider leaving this open with label Documentation.
I think if you close it, there will be duplication.
Thanks.

**Comment by kbernhagen** [comment]
I am attempting to keep up-to-date info on FoldingForum:
[HOWTO: Allow v8 Client Remote Control](https://foldingforum.org/viewtopic.php?t=39050)
[HOWTO: Local v8 Web Control](https://foldingforum.org/viewtopic.php?t=39131)
